initial spec for daemon synchronization (#2035)
thin-edge/thin-edge.io: docs/src/references/synchronize-daemons.md

 For example, when the tedge-mapper-c8y comes up, it will try to get the jwt token. The request will be sent to c8y cloud
 over c8y/s/dat, if the bridge is not up then the request message will be lost.
 So, better to check the status of the bridge before sending the request to the c8y cloud.

3. Some of the daemons are dependent on the `tedge-mapper-c8y`(cloud mapper) file system to create the supported operations file.

# Proposed solutions and their Pros and Cons

## Using the `health check` mechanism to synchronize `tedge-mapper-c8y` and `tedge-agent`

- The `tedge-mapper-c8y` can use the health-check mechanism to detect the health of the `tedge-agent` and then send the request to the `tedge-agent`.
- When the mapper starts it will subscribe to the `tedge/health/tedge-agent` topic.
- When the `tedge-agent` starts, it will publish a `health` status (up/down) message on the `tedge/health/tedge-agent` topic.
- Based on this `tedge-agent` health status message, the mapper can create the supported operations in the `/etc/tedge/operations/c8y` directory.
- Mapper will send the supported operations list to the c8y cloud.

> - Mapper will send the request to the `software list`.

Author: fixed

- Once it receives the software list from `tedge-agent`, the mapper will send the list to the c8y cloud.

> ### Retrieving the internal-id

Author: Good idea, I have updated this section.

> When the tedge-mapper-c8y starts, it will request the `internal id` of the `thin-edge` device.

Author: fixed

Which will be used by the mapper for further communication with the c8y cloud over HTTP.

> To get the `internal id`, the mapper needs the `JWT` token, which was retrieved over the `MQTT`.

Author: fixed

Cons: 
-   This leads to hard dependency on the list of daemons in the mapper. For example, the tedge-mapper-c8y has to subscribe to tedge/health/tedge-agent
topic to receive the health status from the tedge-agent. Based on this the mapper has to create a supported operation file in /etc/tedge/operations/c8y.
-   This will be a hard-coded dependency of the name of the plugin and can’t be updated on the fly if someone wants to change the name of the daemon
    or wants to add a new daemon.

    ## Proposal 2:  Introduce an `init message` mechanism

    ```admonish note
    Pre-condition for this proposal is that the `plugins` must be `cloud-agnostic` and the mappers will be `cloud-specific`.
    The plugins will be grouped based on the roles they perform. For example, software-management, device-restart, config-management, log-management, firmware-management,.etc.
    ```

    The tedge plugins must publish a `retained init` message describing their role (software-management, config management, etc)
     and other details specific to that role.

    The proposed topic will be `tedge/<device-id>/init/<cloud>`.

>     Here `<cloud>` component is `optional` and could be c8y, azure, aws, etc

Reviewer: Such a topic *cannot* be use to publish the init state of the plugins. For that we need topics that are each *uniquely* attached to a single plugin, *the* plugin which capability is declared.

Hence, we need something closer to:

```
tedge/<device-id>/init/<capability-name>
```

- This should be designed by https://github.com/thin-edge/thin-edge.io/pull/2030.
- Why the name `init`?
- Why the device id?
- What about child devices?

As a general remark, that applies to most of our proposal for topic schema: one must stop defining topic schema without data model.

    The message template is {Operation_name: <content>, Operation_name:<content>, Configs: type1..typen}.
    Here `Operation_name` is the name of the operation to be published to the cloud and 

>     the `content` is the content of the operation file, if any.

Reviewer 2: By the operation `content`, if you meant things like the contents of `c8y-config-plugin.toml` file, I really wouldn't add those to the init message as that is purely the plugin's concern and the mapper has no use for it. In the case of the config plugin, it only needs to declare the config types/paths that it supports, which is something that the cloud is interested in (to send the supported configs list).

>     Here `Configs` is the other configuration information that is to be passed to the cloud.

Reviewer 2: Having `Configs` as a separate key, disjoint from the `Operation_name` is problematic, especially when a plugin declares multiple capabilities each with their own configs of different kinds. Otherwise you won't be able to know which `Configs` is for which `Operation_name`.

    For Example, in case of `log-plugin` the `Configs` is log types that are supporeted.

    Example 1:

>     - When the tedge-mapper-c8y is started, it will subscribe to `tedge/<device-id>/init/#` topic

Reviewer 2: ```suggestion
    - When the tedge-mapper-c8y is started, it will subscribe to `tedge/<device-id>/init` and `tedge/<device-id>/init/c8y` topics
```
As I don't see why the tedge mapper would be interested in a `#` subscription that includes init messages meant for other clouds as well.

# Introduction
This document explains the scenarios that require the startup of different tedge daemons to be synchronized,

> and solutions to avoid this synchronization by relying on dynamic runtime health status.

Author: Agree, we must synchronize all the processes, and check their status before delegating or requesting some work from the other side.

# Problem statement

Thin-edge.io has three main issues for synchronizing the daemons

1. The thin-edge daemons depend on another daemon's liveness to delegate or request some work.
 For example, tedge-mapper-c8y depends on tedge-agent. When the tedge-mapper-c8y is started,
 the mapper sends a request to get the existing software list on `tedge/commands/req/software/list` topic,

>  if the tedge-agent is not up and running then the request message will be lost. So, it's better to check if the agent

Author: I agree, to make sure the `tedge-agent` was up at least once, we must change the agent to declare the supported operations `device-restart and software` when it starts for the first time, now the tedge-mapper can pick this and send it across to the cloud. This will make the tedge-agent to be in line with the other plugins as well.

 is up and running before sending the request to get the software list.

2. The thin-edge daemons are dependent on the liveness of the bridge to communicate with the cloud.
 For example, when the tedge-mapper-c8y comes up, it will try to get the jwt token. The request will be sent to c8y cloud

>  over c8y/s/dat, if the bridge is not up then the request message will be lost.

Reviewer: > Again, not quite sure if the requirement is that strict, that the bridge must be up for us to send a message. We only need to ensure that the bridge was established at least once before the request is sent. Once the bridge is established, messages can be sent which will stay with the persistent session of the broker, and get forwarded to the cloud eventually, when the bridge comes up.

Things are a bit more complex here as we need to distinguish the cases where:
- the bridge has been *defined* (using `tedge connect c8y`),
- the bridge is actually *up* (i.e connected to c8y).

If the bridge is defined but down for some reason, then you are correct: the broker will persist the messages sent to c8y. However, if the bridge is not defined (someone run `tedge disconnect c8y`) then all the messages sent to `c8y/#` are simply discarded by the broker as there is no subscribers.

>  So, better to check the status of the bridge before sending the request to the c8y cloud.

Author: > > when the tedge-mapper-c8y comes up, it will try to get the jwt token
> 
> Yes, but this is not the main point: what the c8y mapper needs is the c8y internal id of the device and this internal id is itself only required to use the c8y REST API (say to send a large event which payload doesn't fit in an MQTT message).

Agreed, but the `Internal ID` is retrieved only once and this is re-used throughout the mapper life cycle right?.

> 
> > So, better to check the status of the bridge before sending the request to the c8y cloud.
> 
> Yes, definitely. For instance, sending the list of software must be done not when the mapper start but each time the bridge to c8y is established. A bit more complex, because the agent must also be up.

The liveness of the `c8y` bridge and `tedge-agent` is not checked at the same time, before forwarding the request to the tedge-agent the liveness will be checked, and the `c8y` bridge liveness is required only when the mapper has to communicate with the cloud.
 
> Unfortunately, things are a bit more complex when the bridge is down: some capabilities are no more available. This is not an issue when the requests are triggered from the cloud (as for the operations). It's an issue when the mapper uses HTTP instead of MQTT (e.g for large events - because the events sent over HTTP will be lost when the bridge is down and not persisted as for those sent over MQTT). This issue is far beyond the scope of this proposal.

The whole proposal is only for the MQTT base communication.

>  So, better to check the status of the bridge before sending the request to the c8y cloud.

Reviewer: > Agreed, but the Internal ID is retrieved only once and this is re-used throughout the mapper life cycle right?.

Yes. However, getting this internal id on start is where all the synchronization issues have been observed.

- Based on this `tedge-agent` health status message, the mapper can create the supported operations in the `/etc/tedge/operations/c8y` directory.
- Check if the `c8y-bridge` is up before sending the init messages, including `supported operations` list.
- Mapper will send the supported operations list to the c8y cloud.
- Mapper will send the request `software list` to `tedge-agent`.
- Once it receives a software list from `tedge-agent`, the mapper will send the list to the c8y cloud.

### Communicating with Cumulocity over HTTP

The `tedge-mapper-c8y` needs `internal-id` of the `thin-edge` device from Cumulocity to communicate over HTTP.
When the tedge-mapper-c8y starts, it will request the `internal id` of the `thin-edge` device from cumulocity cloud.
Which will be used by the mapper for further communication with the c8y cloud over HTTP.

To get the `internal id` over HTTP, the mapper needs the `JWT` token, which was retrieved over the `MQTT`.
So, before sending the request to retrieve the `JWT` token on the `c8y/s/uat` topic,
the mapper must check the health of the `c8y bridge`.
When the bridge is up, it can send a request for the JWT token.

> Then the mapper will receive the `JWT` token on the `c8y/s/dat` topic.

Reviewer: > So, before sending the request to retrieve the `JWT` token on the `c8y/s/uat` topic,
the mapper must check the health of the `c8y bridge`.

This writing gives a hint of the problem but is quite far of how things can work. The mapper *reacts* on events telling the bridge is up and takes these events as an opportunity to fetch the internal device id if not already done.

- Now the mapper picks up this message and creates the supported operations file (c8y_SoftwareUpdate, etc) in `/etc/tedge/operations/c8y` with their content. 
- Mapper sends the updated supported operations list to the c8y cloud.
- Also, now that the tedge-agent is up and running, the tedge-mapper-c8y will publish the request to get the
 software list on `tedge/commands/req/software/list`.
- Now the tedge-agent processes this request and sends the software list to the mapper, once the mapper gets the response it will forward it to the c8y cloud.

Example 2,
- When tedge-mapper-c8y is started, it will subscribe to `tedge/health/#` topic
- When c8y-log-plugin is started, it will publish health status as `up` on to `tedge/health/c8y-log-plugin` topic
- Now the mapper picks up this message and creates the c8y_LogfileRequest operation file in `/etc/tedge/operations/c8y`.
- Also the mapper creates the `c8y-log-plugin.toml` file, which contains the `supported log types` in `/etc/tedge/c8y`.

> - Mapper sends the updated supported operations list to the c8y cloud. Also, it sends the supported `log types`(118,software-management) to the c8y cloud.

Reviewer: > Because when the log-plugin restarts, this will emit the health status again,

This is what I meant by flaky: depending a plugin is running or not the operation will available or not in the cloud.

I think that:
- A capability must be displayed as available to the user when the associated plugin is installed.
- If the plugin is not running, stopped or broken, then the capability is not operational but still available (i.e. listed among the things the device is supposed ready to do).

> - Mapper sends the updated supported operations list to the c8y cloud. Also, it sends the supported `log types`(118,software-management) to the c8y cloud.

Author: The mapper here need not create the `c8y-log-plugin.toml` as it's of no use to the mapper. Only it needs these types(config or log), to send to the cloud.

If there is any issue sending the operation status and list, it must check the reason for the error and,
If the failure is due to the `internal-id` then it must get the internal id again and must re-try sending the operation result.

> 

Reviewer 2: Even though caching of the `jwt` token and `internal-id` are not mentioned in this spec, I hope your impl would be doing that anyway. In-memory caching to start with. We can build the persistance bits later on.

Cons: 
-   This leads to hard dependency on the list of daemons in the mapper. For example, the tedge-mapper-c8y has to subscribe to tedge/health/tedge-agent
topic to receive the health status from the tedge-agent. Based on this the mapper has to create a supported operation file in /etc/tedge/operations/c8y.
-   This will be a hard-coded dependency of the name of the plugin and can’t be updated on the fly if someone wants to change the name of the daemon
    or wants to add a new daemon.

## Proposal 2:  Introduce an `init message` mechanism

Make all the tedge components publish a `retained init` message describing their role (software management, config management, etc)
 and other details specific to that role.

> The proposed topic will be `tedge/<device-id>/init/<cloud>`.

Reviewer 2: Although I understand the need for cloud specific init endpoints(so that devs can develop cloud specific plugins as well), we should also provide cloud agnostic init endpoints like just tedge/init to which cloud agnostic plugins can declare their capabilities in a cloud agnostic manner. If we only support cloud specific init endpoints, we'll be forcing all the plugins also to be cloud aware, which is against our cloud-agnostic principle.

Author: Yes, I changed the spec. Now the `cloud-specific` mapper will subscribe to both the `tedge/<device-id>/init` as well as `tedge/<device-id>/init/<cloud>` both topics. The plugins have the freedom to send the init message either to the `cloud-specific` or to the `generic` init topic.

> Here `<cloud>` could be c8y, azure, aws, etc

Author: In certain cases, if a plugin is specific to the cloud then it can publish the init message only for that `cloud` specific topic. Then, the cloud-specific mapper can pick this up and create the supported operation in the specific operations directory and forward the supported operation to that specific cloud.

Also, when the same device has multiple different cloud mappers running and each cloud supports only specific plugins then this is helpful.

So, the cloud-specific mapper will subscribe to both generic `tedge/<device-id>/init` and cloud-specific `/tedge/<device-id>/init/<cloud>` topics and perform the required operation when it receives the message on those topics.

> Here `<cloud>` could be c8y, azure, aws, etc

Reviewer 2: It's not clear from the above text that the primary init topic is just `tedge/<device-id>/init and the `<cloud>` part is optional. Better to make that explicit.

> Here `<cloud>` could be c8y, azure, aws, etc

Reviewer: > Also, when the same device has multiple different cloud mappers running and each cloud supports only specific plugins then this is helpful.

This is the reverse, making things more complex. Indeed, a plugin should not have to know which cloud will support the associated operation. While, a mapper can subscribe to all plugins' init messages and picks only those actually supported.

The message template is {Operation_name: <content>, Operation_name:<content>, Config_types: type1..typen}.
Here `Operation_name` is the name of the operation to be published to the cloud and 
the `content` is the content of the operation file, if any.

> Here `Config_types` is the configuration types that are supported by that particular daemon.

Reviewer 2: > Its still ok the c8y-configuration-plugin in its init message can send `{DownloadConfigFile:, UploadConfigFile:<>, Config_types: "type1..typen"}

Just for a very short term, as we've already planned to make those `c8y-xyz-plugin`s cloud agnsotic. So, we really need to provide a cloud agnostic interface for them to declare their capabilities and details of those capabilities(like config types list for config-plugin and log types for log-plugin etc).

Example 1:
- When the tedge-mapper-c8y is started, it will subscribe to `tedge/<device-id>/init/#` topic

> - When tedge-agent is started, it will publish `{}` an empty `init` message on to `tedge/init/c8y/tedge-agent`

Reviewer 2: As mentioned earlier, `tedge-agent` which is a cloud agnsotic component, must not publish to the cloud specific`tedge/init/c8y` topic. And it must not be an empty message either. The `tedge-agent` must explicitly declare that it supports `software_list`, `software_update` and `device_restart` operations, which the mapper can map to `c8y_SoftwareUpdate`, `c8y_SoftwareList` and `c8y_Restart` operations.

Author: Yes, agreed. Updated spec.

> - Now the mapper picks up this message and creates the `c8y_Restart, c8y_SoftwareUpdat` supported operations file in `/etc/tedge/operations/c8y`. 

Reviewer 2: ```suggestion
- Now the mapper picks up this message and creates the `c8y_Restart, c8y_SoftwareUpdate` supported operations file in `/etc/tedge/operations/c8y`. 
```

Author: fixed

- Mapper sends the updated supported operations list to the c8y cloud.
- Also, now that the tedge-agent is up and running, the tedge-mapper-c8y will publish the request to get the

>  software list on `tedge/commands/req/software/list`.

Reviewer: > When tedge-agent is started, it will publish `{}` an empty `init` message on to `tedge/init/c8y/tedge-agent`

If this message is empty, what the difference with a health status?

> Now the mapper picks up this message and creates the `c8y_Restart, c8y_SoftwareUpdat` supported operations file in `/etc/tedge/operations/c8y`.

This is okay the mapper has to know the mapping from tedge capabilities to c8y capability.

> Mapper sends the updated supported operations list to the c8y cloud.

We need to be sure that the operations published on the cloud are not  added/removed each time the agent is started/stopped. I guess the intent is to have these capabilities published/removed on install/uninstall on this new topic.

> Also, now that the tedge-agent is up and running, the tedge-mapper-c8y will publish the request to get the
 software list on `tedge/commands/req/software/list`.

This is where the proposed approach to capability discovery falls short. The mapper *still* has to know that the `c8y_SoftwareUpdate` capability is provided via  the `tedge/commands/req/software/list` topic. Somehow the mapper gains no more information with a init capability message than with a simple name as 'tegde-agent' is installed.

Being aware that a capability is available is not enough. The mapper needs to know how to use it, to invoke the capability and to translate the messages. A well-known name is enough to trigger the associated code.

Author: > > When tedge-agent is started, it will publish `{}` an empty `init` message on to `tedge/init/c8y/tedge-agent`
> 
> If this message is empty, what the difference with a health status?

Yes, this is only for the `tedge-agent` case. Because it will not send any capabilities.
But this can be changed later and ask the `tedge-agent` to publish the capabilities.

> 
> > Now the mapper picks up this message and creates the `c8y_Restart, c8y_SoftwareUpdat` supported operations file in `/etc/tedge/operations/c8y`.
> 
> This is okay the mapper has to know the mapping from tedge capabilities to c8y capability.
> 
> > Mapper sends the updated supported operations list to the c8y cloud.
> 
> We need to be sure that the operations published on the cloud are not added/removed each time the agent is started/stopped. I guess the intent is to have these capabilities published/removed on install/uninstall on this new topic.

If we just stick to publishing the capabilities on `install/uninstall` then the dependency is MQTT broker must be running during installation.

So, I feel we have to make a trade-off here, even if the agent sends the capabilities when it starts. The c8y mapper will try creating and publishing the same set of capabilities.

> 
> > Also, now that the tedge-agent is up and running, the tedge-mapper-c8y will publish the request to get the
> > software list on `tedge/commands/req/software/list`.
> 
> This is where the proposed approach to capability discovery falls short. The mapper _still_ has to know that the `c8y_SoftwareUpdate` capability is provided via the `tedge/commands/req/software/list` topic. Somehow the mapper gains no more information with a init capability message than with a simple name as 'tegde-agent' is installed.

Agree, the plan here is to still keep the mapper cloud-specific. So, it will continue to do the same thing as it is doing as of now except that creating capabilities on the right directories and publishing those capabilities to the cloud.
The main advantage here is the `plugin` need not worry about creating the `operations` file. This will be helpful when each plugin is containerized and run in different containers.
> 
> Being aware that a capability is available is not enough. The mapper needs to know how to use it, to invoke the capability, and to translate the messages. A well-known name is enough to trigger the associated code.

Reviewer: >> > When tedge-agent is started, it will publish `{}` an empty `init` message on to `tedge/init/c8y/tedge-agent`
>> 
>> If this message is empty, what the difference with a health status?
>
> Yes, this is only for the `tedge-agent` case. Because it will not send any capabilities.
> But this can be changed later and ask the `tedge-agent` to publish the capabilities.

The fact that things can work with an empty payload is an argument for my point: a capability is captured with a well-known name that *implies* a set of operations, message schema and API calls.

In this example, the mapper has to know that `tedge-agent` is a daemon that can do two things: restart the device and install software. By the way, it would be good to distinguish the daemon from the capabilities. Here, it would be better to have two: `device-restart` and `software-update`.

- Now the tedge-agent processes this request and sends the software list to the mapper, once the mapper gets the response it will forward it to the c8y cloud.

Example 2,
- When tedge-mapper-c8y is started, it will subscribe to `tedge/<device-id>/init/#` topic

> - When c8y-log-plugin is started, it will publish {"LogfileRequest":"", "LogTypes": "software-management, mosquitto-log"} on to `tedge/thin-edge/init` topic

Reviewer 2: ```suggestion
- When c8y-log-plugin is started, it will publish {"c8y_LogfileRequest":"", "c8y_SupportedLogs": "software-management, mosquitto-log"} on to `tedge/thin-edge/init` topic
```
If we're supporting cloud specific endpoints anyway, then I'd keep the keys as close to the cloud supported ones so that the developer doesn't need to learn an extra one.  Sending the operation type is still relevant as the mapper needs to combine that with other supported operations as well. But, we can even skip keys like `c8y_SupportedLogs` as there is no benefit in making the mapper send them. The plugin could send it directly.

Author: If `plugin` has to send `c8y_SupportedLogs` then the plugin must be aware of the cloud. But here, we want to make the `plugins` cloud agnostic.

Reviewer 3: Ideally the `init` command should be as cloud agnostic as possible, so I would also be in favour of the payload not having cloud specific names inside it. We can still use the Cumulocity IoT contract as a good reference point since the other Clouds don't offer these features out of the box.

Reviewer 2: Oh don't get me wrong. I also want all our plugins to be as cloud-agnostic as possible. I was saying that it'd be better to keep the keys similar to that of the cloud **only** for the cloud specific init topics. If the `c8y-log-plugin` is converted into a cloud agnostic `tedge-log-plugin`, we wouldn't need it.

- Now the mapper picks up this message and creates the c8y_LogfileRequest operation file in `/etc/tedge/operations/c8y`.
- Also, the mapper creates the `c8y-log-plugin.toml` file, which contains the `supported log types` in `/etc/tedge/c8y`.

> - Mapper sends the updated supported operations list to the c8y cloud. Also, it sends the supported `log types`(118,software-management, mosquitto-log) to the c8y cloud.

Reviewer: Aren't these points under the responsibility of the log plugin?

Author: Yes, as of now it is the `c8y-log-plugin's` responsibility. But when it is translated to the `tedge-log-plugin` then it makes sense to move all the cloud-specific things to the cloud-specific mapper.

> Note: The mapper can also subscribe to cloud-specific init topics like `tedge/<device-id>/init/c8y`, to get a init message from a cloud-specific plugin,
that does not belong to any of the generic operations categories. For example, any custom operation plugin in case of c8y cloud.

Pros:
-   This will remove the file system dependency when the daemons want to create the supported operation files on the tedge-mapper-c8y file system.
-	No dependency on mapper, no hard-coded list of plugins
-	Very much cloud agnostic because the plugins are categorized based on roles not cloud-specific.

>  For example, the `log` plugin can be used by c8y, azure, aws, etc.

Reviewer: This is why I don't understand why the log plugin sends its init message on a topic that is specific to a cloud
(i.e. `tedge/<device-id>/init/<cloud>`)

Author: Yes, in that case, it need not send on a cloud-specific topic.

 Based on the mapper type, the operations will be created in those specific directories and the supported operations list will be forwarded to the particular cloud.
    
Cons:

> -   Need one more topic, and one more message format, which might be challenging to maintain.

Reviewer: This is not a concern as long as there is benefit.

# Shutting down or removing the plugin permanently
 

>  When a plugin is shut down gracefully and removed from active operation, the operation must be removed from the list of operations supported by that 

Reviewer 2: I wouldn't do that on every shutdown of the plugin but only when it's really uninstalled. Else we'd be removing and adding ops to a device quite a lot every time a plugin is "gracefully" restarted.
